### PR TITLE
Add continuation state API

### PR DESCRIPTION
**Pull request by andimarafioti**
## What changed

This adds a caller-owned continuation-state API for follow-up turns across all public generation modes:

- `generate_voice_clone`
- `generate_voice_clone_streaming`
- `generate_custom_voice`
- `generate_custom_voice_streaming`
- `generate_voice_design`
- `generate_voice_design_streaming`

The generators can now return a full continuation state on the first turn, return delta-only state updates on later turns, and accept a previously returned full state to continue from existing talker cache state instead of restarting cold.

The implementation also carries decoder left-context forward so chunk/audio boundaries do not decode with a cold codec context, and it adds `continuation_state_status` metadata so callers can warn or reset before hitting `max_seq_len`.

## Why it changed

When an application synthesizes multi-sentence LLM output incrementally, restarting TTS from scratch on each sentence makes the follow-up sentence sound detached from the previous one. The goal here is to let callers preserve enough internal state to continue naturally across turns while keeping the state caller-owned so servers can multiplex or persist it externally.

## Impact

- Improves continuity across separately generated turns
- Works for Base voice clone, CustomVoice, and VoiceDesign
- Works for both streaming and non-streaming audio output
- Lets callers merge only newly generated cache rows with `apply_continuation_state_delta(...)`
- Adds `benchmarks/continuation.py` to measure fresh vs continuation follow-up latency
- Documents the new API in the README

## Validation

- `./.venv/bin/pytest tests/test_continuation_state_api.py tests/test_voice_clone_prompt_api.py -q`
- `./.venv/bin/python -m compileall faster_qwen3_tts`
- GPU smoke tests for `custom_voice`, `voice_clone`, and `voice_design` in streaming and non-streaming continuation paths
- Quick benchmark sanity checks on this machine:
  - `custom_voice`: TTFA `419 ms -> 317 ms`, non-stream latency `2726 ms -> 2421 ms`
  - `voice_clone`: TTFA `381 ms -> 312 ms`, non-stream latency `1893 ms -> 2000 ms`
  - `voice_design`: TTFA `456 ms -> 402 ms`, non-stream latency `2077 ms -> 2077 ms`

## Notes

There is still no cache compaction/sliding-window truncation for long continuation sessions. Instead, the API now returns `continuation_state_status` so callers can decide when to start a fresh session before the saved state gets too close to `max_seq_len`.

### CODE REVIEW

**Comment by andimarafioti** [comment]
claude's review:
  PR #103 Review                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                
  Bugs                                                                                                                                                                                                                                          
                                                                                                                                                                                                                                                
  1. parity_generate_streaming is broken — unbound name + bad call-site kwarg                                                                                                                                                                   
                                                                            
  continuation_max_seq_len is referenced inside parity_generate_streaming (streaming.py:374, 375, 382, 419, 588, 632) but it is not in that function's signature (streaming.py:333-353). It exists only on fast_generate_streaming              
  (streaming.py:51). Also, model.py:1038 passes it as a kwarg only in the parity branch — which would TypeError on dispatch:
                                                                                                                                                                                                                                                
  # model.py:1034-1038                                                      
  if not parity_mode:                                                                                                                                                                                                                           
      stream_kwargs["predictor_graph"] = self.predictor_graph
      stream_kwargs["talker_graph"] = self.talker_graph                                                                                                                                                                                         
  else:                                                                                                                                                                                                                                         
      stream_kwargs["continuation_max_seq_len"] = self.max_seq_len
                                                                                                                                                                                                                                                
  Any call to generate_voice_clone_streaming(..., parity_mode=True) raises TypeError, and if that somehow succeeded, the first reference would raise NameError. No test covers this, which is why CI is green. Add continuation_max_seq_len: int
   | None = None to the parity signature and add at least one smoke test.                                                                                                                                                                       
                                                                                                                                                                                                                                                
  Simplification — high value                                                                                                                                                                                                                   
  
  2. Three near-identical continuation-prefill blocks                                                                                                                                                                                           
                                                                            
  Nearly identical 45-line blocks in generate.py:155-201, streaming.py:113-159, streaming.py:416-462 build required_len, build a full attention mask, build prefix_cache_position and prefix_input_ids, set talker.rope_deltas, build the       
  dynamic cache from state, and call talker.forward. Extract to one helper:
                                                                                                                                                                                                                                                
  def _prefill_with_continuation(talker, continuation_state, talker_input_embeds, attention_mask,
                                  trailing_text_hiddens, tts_pad_embed, max_seq_len, device):                                                                                                                                                   
      ...  # returns (out, full_attention_mask, base_seq_len)                                                                                                                                                                                   
                                                                                                                                                                                                                                                
  3. Four near-identical delta-emit blocks                                                                                                                                                                                                      
                                                                                                                                                                                                                                                
  generate.py:302-326, streaming.py:260-283, 305-327, 567-590, 611-633 all build a delta and stamp continuation_state_delta / continuation_state / continuation_state_status into a timing dict. A single helper _emit_continuation(timing,     
  static_cache, base, seq_len, ...) collapses ~25×4 lines.
                                                                                                                                                                                                                                                
  4. Three near-identical non-streaming post-processing blocks              

  generate_voice_clone (model.py:869-910), generate_custom_voice (1183-1219), generate_voice_design (1420-1456) each run the same sequence: resolve decoder_prefix_codes, concat to codec_ids, speech_tokenizer.decode, trim ref portion from   
  each audio array, compute RTF, log, return 2- or 3-tuple. Extract _decode_codec_ids_and_finalize(codec_ids, continuation_state, timing, ref_codes=None).
                                                                                                                                                                                                                                                
  5. Three near-identical streaming post-processing loops                   

  generate_voice_clone_streaming (model.py:1046-1107), generate_custom_voice_streaming (1269-1349), generate_voice_design_streaming (1499-1581) each repeat the calibrate-then-sliding-window codec→audio logic (~60 lines each). Only          
  generate_voice_clone_streaming has the ref_codes path; the other two could share a single helper with ref_codes=None.
                                                                                                                                                                                                                                                
  Simplification — minor                                                                                                                                                                                                                        
  
  6. Duplicated suppress-mask construction                                                                                                                                                                                                      
                                                                            
  generate.py:61-65, streaming.py:64-68, streaming.py:367-371 build the same suppress_mask the same way. Move to sampling.py as build_suppress_mask(vocab_size, eos_id, device).                                                                
  
  7. _build_generation_info explicit field copy                                                                                                                                                                                                 
                                                                            
  model.py:103-108 enumerates three keys individually:                                                                                                                                                                                          
  if "continuation_state_delta" in timing: info["continuation_state_delta"] = timing["continuation_state_delta"]
  if "continuation_state" in timing: ...                                                                                                                                                                                                        
  if "continuation_state_status" in timing: ...                                                                                                                                                                                                 
  Trivially a loop:                            
  for k in ("continuation_state_delta", "continuation_state", "continuation_state_status"):                                                                                                                                                     
      if k in timing: info[k] = timing[k]                                                                                                                                                                                                       
  
  8. bool() around already-bool inputs                                                                                                                                                                                                          
                                                                            
  generate.py:80, streaming.py:83, 392 — continuation_state["non_streaming_mode"] != bool(continuation_non_streaming_mode) — continuation_non_streaming_mode is already normalized to bool by _resolve_non_streaming_mode. Harmless but noise.  
                                                                            
  9. Inconsistent "empty decoder context" representation                                                                                                                                                                                        
                                                                            
  continuation_state_decoder_context returns None when empty (continuation.py:119-121), but _decoder_context_from_delta returns torch.empty(0, 0, dtype=torch.long) when empty (continuation.py:199-201). Downstream readers handle both, but   
  returning None everywhere would be cleaner.                               
                                                                                                                                                                                                                                                
  API / contracts                                                                                                                                                                                                                               
  
  10. Return-tuple shape depends on a kwarg                                                                                                                                                                                                     
                                                                            
  generate_voice_clone etc. return a 2-tuple when return_continuation_state=False and a 3-tuple otherwise (e.g. model.py:909 vs 910). Callers are forced into if len(result) == 3: (see benchmarks/continuation.py:49-52). Either always return 
  the info dict (None when unused) or return a dataclass. Not a blocker since it preserves backwards compatibility, but worth flagging.
                                                                                                                                                                                                                                                
  11. Quadratic cache growth across turns                                                                                                                                                                                                       
  
  Every call to apply_continuation_state_delta calls torch.cat on each layer's K/V (continuation.py:244-249). For long multi-turn sessions this is O(n²) in memory copies. The README hints callers should reset — fine — but consider          
  pre-allocating to a configurable capacity on first full build.            
                                                                                                                                                                                                                                                
  12. Test coverage gaps                                                                                                                                                                                                                        
  
  - No test exercises parity_generate_streaming with continuation — that is why bug #1 is latent.                                                                                                                                               
  - test_apply_continuation_state_delta_builds_and_extends_state does not assert rejection when delta.base_seq_len != state.seq_len nor when version/mode mismatch.
                                                                                                                                                                                                                                                
  Nits                                                                      
                                                                                                                                                                                                                                                
  - generate.py:316/318 — dict literal key ordering is fine but timing["continuation_state_delta"] = delta duplicates the "continuation_state" in info vs "continuation_state_delta" in info branching already present in                       
  _build_generation_info. If you adopt simplification #3 these become one site.
  - CONTINUATION_WARNING_MIN_TOKENS = 64 in continuation.py:10 and DECODER_CONTEXT_FRAMES = 25 are magic numbers; the README documents the 25-frame context in generate_voice_clone_streaming but not the // 8 threshold in                     
  build_continuation_state_status. One short comment on the heuristic would help.

**Comment by andimarafioti** [comment]
Codex's review:

Reviewed PR #103, Add continuation state API (draft). (github.com (https://github.com/andimarafioti/faster-qwen3-tts/pull/103))

  1. High: generate_voice_clone_streaming(..., parity_mode=True) is broken by the new continuation plumbing. /tmp/faster-qwen3-tts-pr103/faster_qwen3_tts/model.py:1012 always injects continuation_max_seq_len, but /tmp/faster-qwen3-tts-
     pr103/faster_qwen3_tts/streaming.py:333 neither accepts that keyword nor defines it before use at line 374. In a direct repro, next(model.generate_voice_clone_streaming(..., parity_mode=True)) raises TypeError:
     parity_generate_streaming() got an unexpected keyword argument 'continuation_max_seq_len'. This is a public regression of an existing option.
  2. High: apply_continuation_state_delta is not device-safe. In /tmp/faster-qwen3-tts-pr103/faster_qwen3_tts/continuation.py:241 the cache and first-token history are moved onto the existing state's device, but /tmp/faster-qwen3-tts-pr103/
     faster_qwen3_tts/continuation.py:260 concatenates codec_ids_delta without a .to(...), and /tmp/faster-qwen3-tts-pr103/faster_qwen3_tts/continuation.py:271 clones rope_deltas without normalization. A CPU full state plus a GPU delta
     reproduces a RuntimeError about mixed devices. Given the new continuation_state_device API, that mismatch is plausible in real use.
  3. Medium: apply_continuation_state_delta(None, delta) accepts deltas whose base_seq_len is already non-zero and silently fabricates a corrupt full state. The state is None branch in /tmp/faster-qwen3-tts-pr103/faster_qwen3_tts/
     continuation.py:220 only clones cache_delta, but the returned object still records seq_len = delta["seq_len"] in /tmp/faster-qwen3-tts-pr103/faster_qwen3_tts/continuation.py:269. Minimal repro: a delta with base_seq_len=3 and
     added_seq_len=2 returns seq_len=5 while the cache only contains 2 positions. That invalid state also passes validate_full_continuation_state, so the failure gets deferred into later continuation calls.

  Targeted validation on the PR checkout passed: tests/test_continuation_state_api.py and tests/test_voice_clone_prompt_api.py (32 passed), and python -m compileall faster_qwen3_tts also passed. The new coverage does not exercise
  parity_mode or mixed-device delta merges, which is where the two runtime issues above slipped through.

**Comment by andimarafioti** [comment]
Follow-up handoff note before switching back to `main`-based test-fix work.

Latest pushed commit: `e739852` (`fix continuation wrapper defaults`).

What this follow-up fixed:
- Accepted the internal normalized continuation return mode (`"none"`) instead of rejecting it on the second normalization pass.
- Fixed the manual repro where `generate_custom_voice(..., do_sample=False)` failed when `return_continuation_state` was left at its default.
- Added broader wrapper-contract tests for all 6 public generation methods so default no-continuation paths are exercised on cheap CPU tests.

Validation on this branch:
- `./.venv/bin/pytest -q tests/test_continuation_state_api.py tests/test_voice_clone_prompt_api.py` -> `49 passed`
- CUDA subset run: `./.venv/bin/pytest -q tests/test_sampling.py tests/test_continuation_state_api.py tests/test_e2e_parity.py` -> `32 passed, 4 failed`

The remaining 4 failures are still the same pre-existing ones from `main`:
- `tests/test_e2e_parity.py::TestCustomVoice::test_custom_voice_full_parity_dynamic_cache`
- `tests/test_e2e_parity.py::test_instruct_prepends_tokens_to_voice_clone`
- `tests/test_e2e_parity.py::test_instruct_changes_generation_output`
- `tests/test_e2e_parity.py::TestVoiceDesign::test_voice_design_full_parity_dynamic_cache`

I’m leaving continuation work here for now and moving to a fresh `main`-based branch to fix those test failures directly.

**Comment by andimarafioti** [comment]
Merged `main` into this branch and reran the full test suite, including the CUDA coverage.

Current result on `codex/add-continuation-state-api`:
- `./.venv/bin/pytest -q` -> `71 passed, 3 warnings`

So this branch is in a good state now after syncing with `main`.
